src/main/java/com/flightstats/hub/spoke/RemoteClusterSpokeStore.java

```diff
@@ -0,0 +1,356 @@
+                    try {
+                        ContentKey key = new ContentKey();
+                        if (insert(SpokeStore.WRITE, path + key.toUrl(), key.toUrl().getBytes(), server, traces, "payload", path)) {
+                            quorumLatch.countDown();
+                        } else {
+                            traces.log(log);
+                        }
+                    } catch (Exception e) {
+                        log.warn("unexpected exception " + server, e);
+                    }
+                }
+            });
+        }
+        if (quorumLatch.await(5, TimeUnit.SECONDS)) {
+            threadPool.shutdown();
+            log.info("completed warmup calls to Spoke {}", server);
+        } else {
+            threadPool.shutdown();
+            throw new RuntimeException("unable to properly connect to Spoke " + server);
+        }
+    }
+
+    @Override
+    public boolean testAll() throws UnknownHostException {
+        Collection<String> servers = cluster.getRandomServers();
+        servers.addAll(Cluster.getLocalServer());
+        log.info("*********************************************");
+        log.info("testing servers {}", servers);
+        log.info("*********************************************");
+        String path = HubHost.getLocalAddressPort();
+        for (String server : servers) {
+            try {
+                log.info("calling server {} path {}", server, path);
+                String url = HubHost.getScheme() + server + "/internal/spoke/test/" + path;
+                ClientResponse response = query_client.resource(url).get(ClientResponse.class);
+                if (response.getStatus() == 200) {
+                    log.info("success calling {}", response);
+                } else if (response.getStatus() == 404) {
+                    log.warn("test not yet implemented {}", response);
+                } else {
+                    log.warn("failed response {}", response);
+                    return false;
+                }
+            } catch (Exception e) {
+                log.warn("unable to test " + path + " with " + server, e);
+                return false;
+            }
+        }
+        log.info("all startup tests succeeded  " + path);
+        return true;
+    }
+
+    @Override
+    public boolean insert(SpokeStore spokeStore, String path, byte[] payload, String spokeApi, String channel) {
+        return insert(spokeStore, path, payload, cluster.getWriteServers(), ActiveTraces.getLocal(), spokeApi, channel);
+    }
+
+    @Override
+    public boolean insert(SpokeStore spokeStore, String path, byte[] payload, Collection<String> servers, Traces traces,
+                          String spokeApi, String channel) {
+        int quorum = getQuorum(servers.size());
+        CountDownLatch quorumLatch = new CountDownLatch(quorum);
+        AtomicBoolean firstComplete = new AtomicBoolean();
+        for (final String server : servers) {
```

Comment:
There's some gnarly deeply nested code here...but it looks like it's just moved from a different location, so that's ok. If you were going to be making any actual changes (logic/flow) then I'd suggest some refactoring to extract helper methods and simplify.

Reply:
Yeah, I'm trying to focus this PR on just supporting r/o mode, without doing too much refactoring beyond that.  Definitely worth fixing later though.